File: cocos/cocos2d-x/cocos/platform/android/java/src/org/cocos2dx/lib/Cocos2dxRenderer.java
/****************************************************************************
Copyright (c) 2010-2011 cocos2d-x.org
Copyright (c) 2013-2016 Chukong Technologies Inc.
Copyright (c) 2017-2018 Xiamen Yaji Software Co., Ltd.

http://www.cocos2d-x.org

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.
 ****************************************************************************/
package eggy.cocos2dx.lib;

import android.opengl.GLSurfaceView;

import java.lang.ref.WeakReference;

import javax.microedition.khronos.egl.EGLConfig;
import javax.microedition.khronos.opengles.GL10;

public class Cocos2dxRenderer implements GLSurfaceView.Renderer {
    private final static String TAG = "Cocos2dxRenderer";

    private final static long NANOSECONDSPERSECOND = 1000000000L;
    private final static long NANOSECONDSPERMICROSECOND = 1000000;

    private static final long INTERVAL_60_FPS = (long) (1.0 / 60 * Cocos2dxRenderer.NANOSECONDSPERSECOND);
    private static long sAnimationInterval = INTERVAL_60_FPS;
    private static WeakReference<Cocos2dxRenderer> sRenderer;

    private long mLastTickInNanoSeconds;
    private int mScreenWidth;
    private int mScreenHeight;
    private boolean mNativeInitCompleted = false;
    private boolean mNeedShowFPS = false;
    private String mDefaultResourcePath = "";
    private long mOldNanoTime = 0;
    private long mFrameCount = 0;
    private boolean mNeedToPause = false;

    public static void setPreferredFramesPerSecond(int fps) {
        Cocos2dxRenderer.sAnimationInterval = (long) (1.0 / fps * Cocos2dxRenderer.NANOSECONDSPERSECOND);
    }

    public void setScreenWidthAndHeight(final int surfaceWidth, final int surfaceHeight) {
        this.mScreenWidth = surfaceWidth;
        this.mScreenHeight = surfaceHeight;
    }

    public void setDefaultResourcePath(String path) {
        if (path == null)
            return;
        mDefaultResourcePath = path;
    }

    public void showFPS() {
        mNeedShowFPS = true;
    }

    public interface OnGameEngineInitializedListener {
        void onGameEngineInitialized();
    }

    private OnGameEngineInitializedListener mGameEngineInitializedListener;

    public void setOnGameEngineInitializedListener(OnGameEngineInitializedListener listener) {
        mGameEngineInitializedListener = listener;
    }

    @Override
    public void onSurfaceCreated(final GL10 GL10, final EGLConfig EGLConfig) {
        mNativeInitCompleted = false;
        Cocos2dxRenderer.nativeInit(this.mScreenWidth, this.mScreenHeight, mDefaultResourcePath);
        mOldNanoTime = System.nanoTime();
        this.mLastTickInNanoSeconds = System.nanoTime();
        mNativeInitCompleted = true;
        if (mGameEngineInitializedListener != null) {
            Cocos2dxHelper.getActivity().runOnUiThread(new Runnable() {
                @Override
                public void run() {
                    mGameEngineInitializedListener.onGameEngineInitialized();
                }
            });
        }
    }

    @Override
    public void onSurfaceChanged(final GL10 GL10, final int width, final int height) {
        Cocos2dxRenderer.nativeOnSurfaceChanged(width, height);
    }

    @Override
    public void onDrawFrame(final GL10 gl) {
        if (mNeedToPause)
            return;

        if (mNeedShowFPS) {
            ++mFrameCount;
            long nowFpsTime = System.nanoTime();
            long fpsTimeInterval = nowFpsTime - mOldNanoTime;
            if (fpsTimeInterval > 1000000000L) {
                double frameRate = 1000000000.0 * mFrameCount / fpsTimeInterval;
                Cocos2dxHelper.OnGameInfoUpdatedListener listener = Cocos2dxHelper.getOnGameInfoUpdatedListener();
                if (listener != null) {
                    listener.onFPSUpdated((float) frameRate);
                }
                mFrameCount = 0;
                mOldNanoTime = System.nanoTime();
            }
        }
        /*
         * No need to use algorithm in default(60 FPS) situation,
         * since onDrawFrame() was called by system 60 times per second by default.
         */
        if (sAnimationInterval <= INTERVAL_60_FPS) {
            Cocos2dxRenderer.nativeRender();
        } else {
            final long now = System.nanoTime();
            final long interval = now - this.mLastTickInNanoSeconds;

            if (interval < Cocos2dxRenderer.sAnimationInterval) {
                try {
                    Thread.sleep((Cocos2dxRenderer.sAnimationInterval - interval)
                            / Cocos2dxRenderer.NANOSECONDSPERMICROSECOND);
                } catch (final Exception e) {
                }
            }
            /*
             * Render time MUST be counted in, or the FPS will slower than appointed.
             */
            this.mLastTickInNanoSeconds = System.nanoTime();
            Cocos2dxRenderer.nativeRender();
        }
    }

    private static native void nativeTouchesBegin(final int id, final float x, final float y);

    private static native void nativeTouchesEnd(final int id, final float x, final float y);

    private static native void nativeTouchesMove(final int[] ids, final float[] xs, final float[] ys);

    private static native void nativeTouchesCancel(final int[] ids, final float[] xs, final float[] ys);

    private static native boolean nativeKeyEvent(final int keyCode, boolean isPressed);

    private static native void nativeRender();

    private static native void nativeInit(final int width, final int height, final String resourcePath);

    private static native void nativeOnSurfaceChanged(final int width, final int height);

    private static native void nativeOnPause();

    private static native void nativeOnResume();

    public void setPauseInMainThread(boolean value) {
        mNeedToPause = value;
    }

    public void handleActionDown(final int id, final float x, final float y) {
        if (!mNativeInitCompleted)
            return;

        Cocos2dxRenderer.nativeTouchesBegin(id, x, y);
    }

    public void handleActionUp(final int id, final float x, final float y) {
        if (!mNativeInitCompleted)
            return;

        Cocos2dxRenderer.nativeTouchesEnd(id, x, y);
    }

    public void handleActionCancel(final int[] ids, final float[] xs, final float[] ys) {
        if (!mNativeInitCompleted)
            return;

        Cocos2dxRenderer.nativeTouchesCancel(ids, xs, ys);
    }

    public void handleActionMove(final int[] ids, final float[] xs, final float[] ys) {
        if (!mNativeInitCompleted)
            return;

        Cocos2dxRenderer.nativeTouchesMove(ids, xs, ys);
    }

    public void handleKeyDown(final int keyCode) {
        if (!mNativeInitCompleted)
            return;

        Cocos2dxRenderer.nativeKeyEvent(keyCode, true);
    }

    public void handleKeyUp(final int keyCode) {
        if (!mNativeInitCompleted)
            return;

        Cocos2dxRenderer.nativeKeyEvent(keyCode, false);
    }

    public void handleOnPause() {
        /**
         * onPause may be invoked before onSurfaceCreated,
         * and engine will be initialized correctly after
         * onSurfaceCreated is invoked. Can not invoke any
         * native method before onSurfaceCreated is invoked
         */
        if (!mNativeInitCompleted)
            return;

        Cocos2dxHelper.onEnterBackground();
        Cocos2dxRenderer.nativeOnPause();
    }

    public void handleOnResume() {
        Cocos2dxHelper.onEnterForeground();
        Cocos2dxRenderer.nativeOnResume();
    }

    private static native void nativeInsertText(final String text);

    private static native void nativeDeleteBackward();

    private static native String nativeGetContentText();

    public void handleInsertText(final String text) {
        Cocos2dxRenderer.nativeInsertText(text);
    }

    public void handleDeleteBackward() {
        Cocos2dxRenderer.nativeDeleteBackward();
    }

    public String getContentText() {
        return Cocos2dxRenderer.nativeGetContentText();
    }

}

File: lib-engine/src/main/kotlin/eggy/cocos2dx/lib/Cocos2dxRenderer.kt
/****************************************************************************
 * Copyright (c) 2010-2011 cocos2d-x.org
 * Copyright (c) 2013-2016 Chukong Technologies Inc.
 * Copyright (c) 2017-2018 Xiamen Yaji Software Co., Ltd.
 *
 * http://www.cocos2d-x.org
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */
package eggy.cocos2dx.lib

import android.opengl.GLSurfaceView
import java.lang.ref.WeakReference
import javax.microedition.khronos.egl.EGLConfig
import javax.microedition.khronos.opengles.GL10

class Cocos2dxRenderer : GLSurfaceView.Renderer {
    private var mLastTickInNanoSeconds: Long = 0
    private var mScreenWidth = 0
    private var mScreenHeight = 0
    private var mNativeInitCompleted = false
    private var mNeedShowFPS = false
    private var mDefaultResourcePath = ""
    private var mOldNanoTime: Long = 0
    private var mFrameCount: Long = 0
    private var mNeedToPause = false
    fun setScreenWidthAndHeight(surfaceWidth: Int, surfaceHeight: Int) {
        mScreenWidth = surfaceWidth
        mScreenHeight = surfaceHeight
    }

    fun setDefaultResourcePath(path: String?) {
        if (path == null) return
        mDefaultResourcePath = path
    }

    fun showFPS() {
        mNeedShowFPS = true
    }

    interface OnGameEngineInitializedListener {
        fun onGameEngineInitialized()
    }

    private var mGameEngineInitializedListener: OnGameEngineInitializedListener? = null
    fun setOnGameEngineInitializedListener(listener: OnGameEngineInitializedListener?) {
        mGameEngineInitializedListener = listener
    }

    override fun onSurfaceCreated(GL10: GL10, EGLConfig: EGLConfig) {
        mNativeInitCompleted = false
        nativeInit(mScreenWidth, mScreenHeight, mDefaultResourcePath)
        mOldNanoTime = System.nanoTime()
        mLastTickInNanoSeconds = System.nanoTime()
        mNativeInitCompleted = true
        if (mGameEngineInitializedListener != null) {
            Cocos2dxHelper.activity!!.runOnUiThread { mGameEngineInitializedListener!!.onGameEngineInitialized() }
        }
    }

    override fun onSurfaceChanged(GL10: GL10, width: Int, height: Int) {
        nativeOnSurfaceChanged(width, height)
    }

    override fun onDrawFrame(gl: GL10) {
        if (mNeedToPause) return
        if (mNeedShowFPS) {
            ++mFrameCount
            val nowFpsTime = System.nanoTime()
            val fpsTimeInterval = nowFpsTime - mOldNanoTime
            if (fpsTimeInterval > 1000000000L) {
                val frameRate = 1000000000.0 * mFrameCount / fpsTimeInterval
                val listener = Cocos2dxHelper.onGameInfoUpdatedListener
                listener?.onFPSUpdated(frameRate.toFloat())
                mFrameCount = 0
                mOldNanoTime = System.nanoTime()
            }
        }
        /*
         * No need to use algorithm in default(60 FPS) situation,
         * since onDrawFrame() was called by system 60 times per second by default.
         */if (sAnimationInterval <= INTERVAL_60_FPS) {
            nativeRender()
        } else {
            val now = System.nanoTime()
            val interval = now - mLastTickInNanoSeconds
            if (interval < sAnimationInterval) {
                try {
                    Thread.sleep((sAnimationInterval - interval) / NANOSECONDSPERMICROSECOND)
                } catch (_: Exception) {
                }
            }
            /*
             * Render time MUST be counted in, or the FPS will slower than appointed.
            */mLastTickInNanoSeconds = System.nanoTime()
            nativeRender()
        }
    }

    fun setPauseInMainThread(value: Boolean) {
        mNeedToPause = value
    }

    fun handleActionDown(id: Int, x: Float, y: Float) {
        if (!mNativeInitCompleted) return
        nativeTouchesBegin(id, x, y)
    }

    fun handleActionUp(id: Int, x: Float, y: Float) {
        if (!mNativeInitCompleted) return
        nativeTouchesEnd(id, x, y)
    }

    fun handleActionCancel(ids: IntArray, xs: FloatArray, ys: FloatArray) {
        if (!mNativeInitCompleted) return
        nativeTouchesCancel(ids, xs, ys)
    }

    fun handleActionMove(ids: IntArray, xs: FloatArray, ys: FloatArray) {
        if (!mNativeInitCompleted) return
        nativeTouchesMove(ids, xs, ys)
    }

    fun handleKeyDown(keyCode: Int) {
        if (!mNativeInitCompleted) return
        nativeKeyEvent(keyCode, true)
    }

    fun handleKeyUp(keyCode: Int) {
        if (!mNativeInitCompleted) return
        nativeKeyEvent(keyCode, false)
    }

    fun handleOnPause() {
        /**
         * onPause may be invoked before onSurfaceCreated,
         * and engine will be initialized correctly after
         * onSurfaceCreated is invoked. Can not invoke any
         * native method before onSurfaceCreated is invoked
         */
        if (!mNativeInitCompleted) return
        Cocos2dxHelper.onEnterBackground()
        nativeOnPause()
    }

    fun handleOnResume() {
        Cocos2dxHelper.onEnterForeground()
        nativeOnResume()
    }

    fun handleInsertText(text: String) {
        nativeInsertText(text)
    }

    fun handleDeleteBackward() {
        nativeDeleteBackward()
    }

    val contentText: String
        get() = nativeGetContentText()

    companion object {
        private const val TAG = "Cocos2dxRenderer"
        private const val NANOSECONDSPERSECOND = 1000000000L
        private const val NANOSECONDSPERMICROSECOND: Long = 1000000
        private const val INTERVAL_60_FPS = (1.0 / 60 * NANOSECONDSPERSECOND).toLong()
        private var sAnimationInterval = INTERVAL_60_FPS
        private val sRenderer: WeakReference<Cocos2dxRenderer>? = null

        @JvmStatic
        fun setPreferredFramesPerSecond(fps: Int) {
            sAnimationInterval = (1.0 / fps * NANOSECONDSPERSECOND).toLong()
        }

        @JvmStatic
        private external fun nativeTouchesBegin(id: Int, x: Float, y: Float)

        @JvmStatic
        private external fun nativeTouchesEnd(id: Int, x: Float, y: Float)

        @JvmStatic
        private external fun nativeTouchesMove(ids: IntArray, xs: FloatArray, ys: FloatArray)

        @JvmStatic
        private external fun nativeTouchesCancel(ids: IntArray, xs: FloatArray, ys: FloatArray)

        @JvmStatic
        private external fun nativeKeyEvent(keyCode: Int, isPressed: Boolean): Boolean

        @JvmStatic
        private external fun nativeRender()

        @JvmStatic
        private external fun nativeInit(width: Int, height: Int, resourcePath: String)

        @JvmStatic
        private external fun nativeOnSurfaceChanged(width: Int, height: Int)

        @JvmStatic
        private external fun nativeOnPause()

        @JvmStatic
        private external fun nativeOnResume()

        @JvmStatic
        private external fun nativeInsertText(text: String)

        @JvmStatic
        private external fun nativeDeleteBackward()

        @JvmStatic
        private external fun nativeGetContentText(): String
    }
}